Add sidebar guide (#565)
withastro/starlight: docs/src/content/docs/guides/sidebar.mdx

Learn more about autogenerated sidebars in the [autogenerated groups](#autogenerated-groups) section.

## Basic configuration

The main configuration for the sidebar is located in the [`sidebar` field](/reference/configuration/#sidebar) of your Starlight configuration located in the `astro.config.mjs` file.

A sidebar is primarily composed of two elements:

- [Links](#links)
- [Groups of links](#groups)

Configuring your sidebar is as simple a defining an array of links and groups in the `sidebar` field of your Starlight configuration.

### Links

A link to an internal or external page must be defined by a `label` and a `link` property.

```js
starlight({
	sidebar: [
		// A link to the CSS & Styling guide.
		{ label: 'CSS & Styling', link: '/guides/css-and-tailwind/' },
		// An external link to the Astro website.
		{ label: 'Astro', link: 'https://astro.build/' },
	],
});
```

With the above configuration, the following sidebar will be generated:

<SidebarPreview
	config={[
		{ label: 'CSS & Styling', link: '/guides/css-and-tailwind/' },
		{ label: 'Astro', link: 'https://astro.build/' },
	]}
/>

> 

Reviewer: 🏷️ guide
````suggestion
Optionally, a link can include a `badge` property to display a badge next to the link label.

```js
starlight({
	sidebar: [
		// A link to the CSS & Styling guide.
		{ label: 'CSS & Styling', link: '/guides/css-and-tailwind/', badge: 'New' },
		// An external link to the Astro website.
		{ label: 'Astro', link: 'https://astro.build/' },
	],
});
```

The above configuration will generate the following sidebar:

<SidebarPreview
	config={[
		{ label: 'CSS & Styling', link: '/guides/css-and-tailwind/', badge: 'New' },
		{ label: 'Astro', link: 'https://astro.build/' },
	]}
/>

To further customize the appearance of the badge you can pass an object to the `badge` property containing a `text` and a `variant` property. The `variant` property can be set to `blue`, `green`, `red`, `pink` or `yellow`. By default, the `variant` property is set to `default`, which will uses the accent color of your site.

```js
starlight({
	sidebar: [
		// A link to the CSS & Styling guide.
		{
			label: 'CSS & Styling',
			link: '/guides/css-and-tailwind/',
			badge: { text: 'New', variant: 'green' },
		},
		// An external link to the Astro website.
		{ label: 'Astro', link: 'https://astro.build/' },
	],
});
```

The above configuration will generate the following sidebar:

<SidebarPreview
	config={[
		{
			label: 'CSS & Styling',
			link: '/guides/css-and-tailwind/',
			badge: { text: 'New', variant: 'green' },
		},
		{ label: 'Astro', link: 'https://astro.build/' },
	]}
/>

````

Author: Thanks a lot @kevinzunigacuellar. I made a few changes in my initial PR to ensure badges previews render properly and added your suggestions with a few changes (do not hesitate to tell me if you disagree with any of them):

- I moved the badges part of the guide to their own section. I think they deserve it and it also adds a ToC entry for them so easier to quickly jump/link to them.
- I added an example of the frontmatter badge configuration for autogenerated link items.

It currently looks like this:

![badges](https://github.com/withastro/starlight/assets/494699/732fdc03-4139-4283-b8c6-699247136dc6)

Reviewer: I like it 💜

A group is defined by a `label` string and an `items` array containing other links and/or groups.

```js
starlight({
	sidebar: [
		// A group of links labelled "Guides".
		{
			label: 'Guides',
			items: [
				{ label: 'Components', link: '/guides/components/' },
				{ label: 'Internationalization (i18n)', link: '/guides/i18n/' },
				// A nested group of links.
				{
					label: 'Styling',
					items: [
						{ label: 'CSS', link: '/guides/css-and-tailwind/' },
						{ label: 'Tailwind', link: '/guides/css-and-tailwind/' },
						{ label: 'Shiki', link: '/guides/css-and-tailwind/' },
					],
				},
			],
		},
	],
});
```

The above configuration will generate the following sidebar:

<SidebarPreview
	config={[
		{
			label: 'Guides',
			items: [
				{ label: 'Components', link: '/guides/components/' },
				{ label: 'Internationalization (i18n)', link: '/guides/i18n/' },
				{
					label: 'Styling',
					items: [
						{ label: 'CSS', link: '/guides/css-and-tailwind/' },
						{ label: 'Tailwind', link: '/guides/css-and-tailwind/' },
						{ label: 'Shiki', link: '/guides/css-and-tailwind/' },
					],
				},
			],
		},
	]}
/>

> 
> By combining links and groups, you can create a wide variety of sidebars layouts.

Reviewer 2: Moved this up top so can be deleted:

```suggestion
```

## Badges

Links can also include a `badge` property to display a badge next to the link label.

```js
starlight({
	sidebar: [
		{
			label: 'Guides',
			items: [
				// A link with a "New" badge.
				{
					label: 'Components',
					link: '/guides/components/',
					badge: 'New',
				},
			],
		},
	],
});
```

The above configuration will generate the following sidebar:

<SidebarPreview
	config={[
		{
			label: 'Guides',
			items: [
				{
					label: 'Components',
					link: '/guides/components/',
					badge: { text: 'New', variant: 'default' },
				},
			],
		},
	]}
/>

> The `badge` property includes a `text` property (the content to display in the badge) and optionally a `variant` property (to define the type of badge shown).

Reviewer 2: > - Is it obvious enough that the badge property can accept a string or an object?

I think this needs clarification here. It’s a bit odd to flow straight from the code example showing `badge: 'New'` to the statement that it can have a `text` property.

Maybe we want a subsection for badge variants? That way we can frame the use case: “Want a variant? Use the object form”. That would also give us space to address your other point of where we can say that the default badge uses the site accent color.

Author: I think this is a good idea. I've added a new section for badge variants which contains the object example and the description of the default badge color and I "tried" to use the wording of some of your previous suggestions.

`badge.text` - a string representing the content to display. (e.g. "New")

`badge.variant` - to override the `default` styling setting, set to one of the following properties for custom styling: `note`, `tip`, `danger`, `caution` or `success`.

```js
starlight({
	sidebar: [
		{
			label: 'Guides',
			items: [
				// A link with a green "New" badge.
				{
					label: 'Components',
					link: '/guides/components/',
					badge: { text: 'Experimental', variant: 'caution' },
				},
			],
		},
	],
});
```

The above configuration will generate the following sidebar:

<SidebarPreview
	config={[
		{
			label: 'Guides',
			items: [
				{
					label: 'Components',
					link: '/guides/components/',
					badge: { text: 'Experimental', variant: 'caution' },
				},
			],
		},
	]}
/>

> ## Autogenerated groups

Author: Totally agree, and it also has the benefit of hiding the fairly long "Customizing autogenerated links in frontmatter" section in the ToC.

> Any group (or sub-group) can be automatically generated, and will be sorted alphabetically by filename. This is helpful when you do not want to manually enter each sidebar item in a group.
> 
> To create an autogenerated group, set `autogenerate` to an object specifying the `directory`. For example, given the following file structure:
> 
> <FileTree>
> 
> - src/
>   - content/
>     - docs/
>       - guides/
>         - components.md
>         - i18n.md
>         - advanced/
>           - project-structure.md
> 
> </FileTree>
> 
> And the following configuration:
> 
> ```js
> starlight({
> 	sidebar: [
> 		{
> 			label: 'Guides',
> 			// Autogenerate a group of links for the 'guides' directory.
> 			autogenerate: { directory: 'guides' },
> 		},
> 	],
> });
> ```

Reviewer 2: Suggesting similar tweaks to above and to flip the order of the configuration and file tree examples. My thinking here is that when someone reads the description of how to configure this, the _most_ important example they see first is the config snippet, so they can associate the text with the code. Then the file tree can follow and the sidebar preview.

````suggestion
Starlight can automatically generate a group in your sidebar based on a directory of your docs.
This is helpful when you do not want to manually enter each sidebar item in a group.
Pages will be sorted alphabetically by filename by default.

Add an autogenerated group using an object with `label` and `autogenerate` properties. Your `autogenerate` configuration must specify the `directory` to use for sidebar entries. For example, with the following configuration:

```js
starlight({
	sidebar: [
		{
			label: 'Guides',
			// Autogenerate a group of links for the 'guides' directory.
			autogenerate: { directory: 'guides' },
		},
	],
});
```

And the following file structure:

<FileTree>

- src/
  - content/
    - docs/
      - guides/
        - components.md
        - i18n.md
        - advanced/
          - project-structure.md

</FileTree>
````

The following sidebar will be generated:

<SidebarPreview
	config={[
		{
			label: 'Guides',
			items: [
				{ label: 'Components', link: '/guides/components/' },
				{ label: 'Internationalization (i18n)', link: '/guides/i18n/' },
				{
					label: 'advanced',
					items: [
						{ label: 'Project Structure', link: '/guides/project-structure/' },
					],
				},
			],
		},
	]}
/>

### Customizing autogenerated links in frontmatter

> Use the [`sidebar` frontmatter field](/reference/frontmatter/#sidebar) to customize autogenerated links.

Reviewer 2: Small suggestion to underline that the frontmatter is not in the Astro config file like most examples so far.

```suggestion
Use the [`sidebar` frontmatter field](/reference/frontmatter/#sidebar) in individual pages to customize autogenerated links.
```

> This field allows you to set a [custom label](/reference/frontmatter/#label) or add a [badge](/reference/frontmatter/#badge) for a page link, to [hide](/reference/frontmatter/#hidden) a link from the sidebar or to define a [custom order](/reference/frontmatter/#order) for all page links.

Reviewer 2: Nit: I try to avoid things like “this” that refer back to previous context as much as possible especially across paragraphs. Also suggesting “sort weighting” vs “order” because although the property is named `order` it doesn’t really set the _order_ necessarily 😅 

```suggestion
Sidebar frontmatter options allow you to set a [custom label](/reference/frontmatter/#label) or add a [badge](/reference/frontmatter/#badge) to a link, [hide](/reference/frontmatter/#hidden) a link from the sidebar, or define a [custom sort weighting](/reference/frontmatter/#order).
```

```md
---
title: My page
sidebar:
  # Set a custom label for the link
  label: Custom sidebar label
  # Show the link in the sidebar

>   hidden: false

Reviewer 2: I’m trying to decide what to do about this one. I don’t love showing this usage as a) it’s not how anyone would ever use this, because `false` is the default and b) it risks suggesting this is _needed_ for the link to show in the sidebar.

Do you think it could be OK to just not show `hidden` in use? Or alternatively we could have _two_ code snippet examples and show `hidden: true` in one that then is not shown in the sidebar preview? We could perhaps even try the code snippets in `<CardGrid />` if we wanted to have them side by side on larger screens? Might look terrible, might not.

Author: This one was already bothering me too but I think I tried to focus too much on trying to get a 1:1 example with the description above. I think we can just remove it considering we have a link to the frontmatter `hidden` option the line above.

(I tried playing with an extra example for it with a new code snippet but it felt a bit too much for me for such a basic and explicit option.)

  # Set a custom order for the link (lower numbers are displayed higher up)
  order: 2
  # Add a badge to the link
  badge:

>     text: Deprecated
>     variant: caution

Reviewer 2: Maybe we use something shorter so it doesn’t wrap in the example sidebar?

```suggestion
    text: New
    variant: tip
```

---
```

> A sidebar including an autogenerated group for the `guides` directory containing a page with the above frontmatter will generate the following sidebar:

Reviewer 2: Small simplification here to match similar styles above.

```suggestion
An autogenerated group including a page with the frontmatter above will generate the following sidebar:
```

<SidebarPreview
	config={[
		{
			label: 'Guides',
			items: [
				{ label: 'A page', link: '#' },
				{
					label: 'Custom sidebar label',
					link: '#',
					badge: { text: 'Deprecated', variant: 'caution' },
				},
				{ label: 'Another page', link: '#' },
			],
		},
	]}
/>

:::note
The `sidebar` frontmatter configuration is only used for autogenerated links and will be ignored for manually defined links.
:::

## Internationalization

> Manually defined link and group labels can be translated using the `translations` property to define a set of labels for each supported language while the `label` property will act as the default locale.

Reviewer 2: I think we can split this into two sentences for slightly more punchiness and clarity:

```suggestion
Use the `translations` property on link and group entries to translate the link or group label for each supported language.
The `label` property will be used for the default locale and for languages without a translation.
```

```js

> starlight({
> 	sidebar: [
> 		{
> 			label: 'Guides',
> 			translations: { es: 'Guías' },
> 			items: [
> 				{
> 					label: 'Components',
> 					translations: { es: 'Componentes' },
> 					link: '/guides/components/',
> 				},
> 				{
> 					label: 'Internationalization (i18n)',
> 					translations: { es: 'Internacionalización (i18n)' },
> 					link: '/guides/i18n/',
> 				},
> 			],
> 		},
> 	],
> });

Reviewer 2: What do you think about breaking the `translations` objects over new lines like this? I think it makes them stand out a bit more although admittedly it also makes the example take up quite a few more lines.

```suggestion
starlight({
	sidebar: [
		{
			label: 'Guides',
			translations: {
				es: 'Guías'
			},
			items: [
				{
					label: 'Components',
					translations: {
						es: 'Componentes'
					},
					link: '/guides/components/',
				},
				{
					label: 'Internationalization (i18n)',
					translations: {
						es: 'Internacionalización (i18n)'
					},
					link: '/guides/i18n/',
				},
			],
		},
	],
});
```

```

Browsing the documentation in Spanish will generate the following sidebar:

<SidebarPreview
	config={[
		{
			label: 'Guías',
			items: [
				{ label: 'Componentes', link: '/guides/components/' },
				{ label: 'Internacionalización (i18n)', link: '/guides/i18n/' },
			],
		},
	]}
/>

## Collapsing groups

Groups of links can be collapsed by default by setting the `collapsed` property to `true`.

```js
starlight({
	sidebar: [
		{
			label: 'Guides',
			// Collapse the group by default.
			collapsed: true,
			items: [
				{ label: 'Components', link: '/guides/components/' },
				{ label: 'Internationalization (i18n)', link: '/guides/i18n/' },
			],
		},
	],
});
```

> The above configuration will generate the following sidebar:

Reviewer 2: Heh. This actually already got the wording more or less exactly that I suggested above! I guess we should flip the `above` so it’s fully consistent:

```suggestion
The configuration above generates the following sidebar:
```

<SidebarPreview
	config={[
		{
			label: 'Guides',
			collapsed: true,
			items: [
				{ label: 'Components', link: '/guides/components/' },
				{ label: 'Internationalization (i18n)', link: '/guides/i18n/' },
			],
		},
	]}
/>

[Autogenerated groups](#autogenerated-groups) respect the `collapsed` value of their parent group:

```js
starlight({
	sidebar: [
		{
			label: 'Guides',
			// Collapse the group and its autogenerated subgroups by default.
			collapsed: true,
			autogenerate: { directory: 'guides' },
		},
	],
});
```

> The sidebar generated by the above configuration will be:

Reviewer 2: Same for consistency:

```suggestion
The configuration above generates the following sidebar:
```

<SidebarPreview
	config={[
		{
			label: 'Guides',
			collapsed: true,
			items: [
				{ label: 'Components', link: '/guides/components/' },
				{ label: 'Internationalization (i18n)', link: '/guides/i18n/' },
				{
					label: 'advanced',
					collapsed: true,
					items: [
						{ label: 'Project Structure', link: '/guides/project-structure/' },
					],
				},
			],
		},
	]}
/>

This behavior can be overridden by defining the `autogenerate.collapsed` property.

```js
starlight({
	sidebar: [
		{
			label: 'Guides',
			// Do not collapse the "Guides" group but collapse its
			// autogenerated subgroups.
			collapsed: false,
			autogenerate: { directory: 'guides', collapsed: true },
		},
	],
});
```

> The generated sidebar will be:

Reviewer 2: ```suggestion
The configuration above generates the following sidebar:
```

---
title: Sidebar Navigation
description: Learn how to set up and customize your Starlight site’s sidebar navigation links.

> ---

Reviewer 2: As it’s a new feature, how do you feel about us highlighting this page for a bit with a badge?

```suggestion
sidebar:
  badge: New
---
```

				{
					label: 'advanced',
					items: [
						{ label: 'Project Structure', link: '/guides/project-structure/' },
					],
				},
			],
		},
	]}
/>

#### Customizing autogenerated links in frontmatter

Use the [`sidebar` frontmatter field](/reference/frontmatter/#sidebar) in individual pages to customize autogenerated links.

Sidebar frontmatter options allow you to set a [custom label](/reference/frontmatter/#label) or add a [badge](/reference/frontmatter/#badge) to a link, [hide](/reference/frontmatter/#hidden) a link from the sidebar, or define a [custom sort weighting](/reference/frontmatter/#order).

```md
---
title: My page
sidebar:
  # Set a custom label for the link
  label: Custom sidebar label
  # Set a custom order for the link (lower numbers are displayed higher up)
  order: 2
  # Add a badge to the link
  badge:
    text: New
    variant: tip
---
```

An autogenerated group including a page with the frontmatter above will generate the following sidebar:

<SidebarPreview
	config={[
		{
			label: 'Guides',
			items: [
				{ label: 'A page', link: '#' },
				{
					label: 'Custom sidebar label',
					link: '#',

> 					badge: { text: 'Deprecated', variant: 'caution' },

Reviewer 2: Oops, I suggested changing the code example but not the preview component here:

```suggestion
					badge: { text: 'New', variant: 'tip' },
```
